### PR TITLE
[sil-module] Create SILFunctionBuilder and hide creation/erasing func…

### PR DESCRIPTION
**Pull request by gottesmm**
…tions on SILModule.

This commit does not modify those APIs or their usage. It just:

1. Moves the APIs onto SILFunctionBuilder and makes SILFunctionBuilder a friend
   of SILModule.
2. Hides the APIs on SILModule so all users need to use SILFunctionBuilder to
   create/destroy functions.

I am doing this in order to allow for adding/removing function notifications to
be enforced via the type system in the SILOptimizer. In the process of finishing
off CallerAnalysis for FSO, I discovered that we were not doing this everywhere
we need to. After considering various other options such as:

1. Verifying after all passes that the notifications were sent correctly and
   asserting. Turned out to be expensive.
2. Putting a callback in SILModule. This would add an unnecessary virtual call.

I realized that by using a builder we can:

1. Enforce that users of SILFunctionBuilder can only construct composed function
   builders by making the composed function builder's friends of
   SILFunctionBuilder (notice I did not use the word subclass, I am talking
   about a pure composition).
2. Refactor a huge amount of code in SILOpt/SILGen that involve function
   creation onto a SILGenFunctionBuilder/SILOptFunctionBuilder struct. Many of
   the SILFunction creation code in question are straight up copies of each
   other with small variations. A builder would be a great way to simplify that
   code.
3. Reduce the size of SILModule.cpp by 25% from ~30k -> ~23k making the whole
   file easier to read.

NOTE: In this commit, I do not hide the constructor of SILFunctionBuilder since
I have not created the derived builder structs yet. Once I have created those in
a subsequent commit, I will hide that constructor.

rdar://42301529

### CODE REVIEW

**Comment by gottesmm** [comment]
@jckarter @slavapestov @rjmccall My plan is to use this to refactor most of the SILFunction creation bloat in SILGen onto a SILGenFunctionBuilder that will only create the function. The old code for emitting the actual SILGen code will remain the same.

Thoughts, concerns, etc?

**Comment by gottesmm** [comment]
Let me elaborate slightly. At minimum my plan is to hide the ability to create SILFunctions onto a SILGenFunctionBuilder that is specific to SILGen. I was just trying to say that the SILGenFunctionBuilder may be a good place to refactor such things.

**Comment by gottesmm** [comment]
@swift-ci smoke test

File: include/swift/SIL/SILFunctionBuilder.h [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+class SILFunctionBuilder {
+  SILModule &mod;
+
+public:
+  SILFunctionBuilder(SILModule &mod) : mod(mod) {}
+  SILFunctionBuilder(const SILFunctionBuilder &) = default;
+  SILFunctionBuilder &operator=(const SILFunctionBuilder &) = default;
+
+  /// \brief Return the declaration of a utility function that can,
+  /// but needn't, be shared between modules.
```
Comment (rjmccall) [inline include/swift/SIL/SILFunctionBuilder.h]: Let's say "different parts of a program" instead of "modules".

**Review by rjmccall — COMMENTED** [review]
Okay.  The idea seems sound.

**Review by atrick — APPROVED** [review]
Looks great to me. I don't know what the constructor limitations are for, but if you get rid of those and fix the top level comment then it's fine.

**Comment by gottesmm** [comment]
I will fix the move only constructor issues. I was hoping to get feedback. Thanks!

**Comment by gottesmm** [comment]
@swift-ci smoke test and merge

**Comment by gottesmm** [comment]
@swift-ci smoke test and merge

**Comment by gottesmm** [comment]
@swift-ci smoke test and merge

**Pull request merged**